content/post/deduplicating_pga_with_apollo.md

```diff
@@ -0,0 +1,387 @@
+That's our very rough estimation of the uniqueness in PGA.
+
+|              | unique at 80%, %  |  unique at 95%, %  |
+|--------------|-------------------|--------------------|
+| Java         | 76                | 99                 |
+| Ruby         | 73                | 97                 |
+| Python       | 63                | 96                 |
+| JavaScript   | 47                | 85                 |
+| Go           | 24                | 65                 |
+| All languages| 54                | 87                 |
+
+## Landing
+
+The final step of the apollo pipeline is the [community detection](https://arxiv.org/abs/1608.00163),
+which gives us the sense of the number of non-similar files in our corpus.
+Consider the [word ladder game](https://en.wikipedia.org/wiki/Word_ladder)
+invented by Lewis Carroll which justifies coding for food:
+
+```
+CODE -> COLE -> COLD -> FOLD -> FOOD
+```
+
+The same way CCs with many files suffer from the "bridges" between densely
+connected clusters of files. Community detection solves this problem nicely
+assigning files to several clusters with different degree of confidence.
+We decided to use the [igraph](http://igraph.org/python/)'s implementation of
+the [Walktrap algorithm](https://www.nature.com/articles/srep30750)
+since it runs reasonably fast and the quality is reasonably good on our data.
+We had two ways to build the graphs on which to run the community detection:
+
+- Include the artificial vertices which represent the buckets, files are
+star-connected to the buckets they were hashed to.
+- Directly connect each pair of files if they were hashed to the same bucket.
+
+While the second method is ideal, it scales quadratically with the number 
+of files in a bucket, whereas the first one scales linearly. Depending
+on the chosen similarity threshold, the first method produces graphs with more
+artificial nodes then the real ones. Therefore we go with
+the first method at the 95% threshold (3 LSH tables) and with the second 
+method at the 80% threshold (9 LSH tables, 3x more buckets).
+
+At the 80% threshold, we could not detect communities in the following large CCs
+because none of the algorithms which we tried ended within one day:
+
+- 261,643 files, of all 5 languages  
+- 123,807 files, all JavaScript
+- 174,916 files, all Java
+- 233,095 files, all Go
+
+We detected 1,270,529 communities at 80% threshold and 671,936 at 95% threshold.
+
+This post continues with the visualizations of some of the connected
+components and the detected communities using [Gephy](https://gephi.org/).
+
+## Soil probes
+
+Of course, we would have to review each of the found groups of similar files
+to ensure that they make sense, which we cannot do for two reasons:
+there are too many files and the duplication criteria are subjective.
+
+We manually labelled 2,000 pairs of Java files as almost the same, similar or
+different using [🐈 src-d/code-annotation](https://github.com/src-d/code-annotation)
```

Comment:
The paragraph above says that this is subjective and so clusters cannot be reviewed and here it's objective enough to be annotable.

I would remove the comment in the paragraph above.

Reply:
why is there a 🐈?

Reply:
![image](https://user-images.githubusercontent.com/2793551/46483014-033d5200-c7f7-11e8-8493-4aca9ae45ebb.png)